util/progress/metricwriter.go

```diff
@@ -149,3 +150,40 @@ func detectLocalSourceType(vertexName string) attribute.KeyValue {
 	// No matches found.
 	return attribute.KeyValue{}
 }
+
+type (
+	execMetricRecorder struct {
+		// Attributes holds the attributes for this metric recorder.
+		Attributes attribute.Set
+
+		// Duration tracks the duration of exec statements.
+		Duration metric.Float64Counter
+	}
+)
+
+func newExecMetricRecorder(meter metric.Meter, attrs attribute.Set) *execMetricRecorder {
+	mr := &execMetricRecorder{
+		Attributes: attrs,
+	}
+	mr.Duration, _ = meter.Float64Counter("exec.command.time",
+		metric.WithDescription("Measures the length of time spent executing run statements."),
+		metric.WithUnit("ms"))
+	return mr
+}
+
+func (mr *execMetricRecorder) Record(ss *client.SolveStatus) {
+	for _, v := range ss.Vertexes {
+		if v.Started == nil || v.Completed == nil || !detectExecType(v.Name) {
+			continue
+		}
+
+		dur := float64(v.Completed.Sub(*v.Started)) / float64(time.Millisecond)
+		mr.Duration.Add(context.Background(), dur, metric.WithAttributeSet(mr.Attributes))
+	}
+}
+
+var reExecType = regexp.MustCompile(`^\[.*] RUN `)
```

Comment:
It feels weird to rely on regex of the progress stream for this.

This will only work with the `dockerfile` frontend, since not all exec ops will necessarily begin with `RUN`.

Reply:
Likewise in https://github.com/docker/buildx/pull/2317, https://github.com/docker/buildx/pull/2271, etc.